include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -0,0 +1,397 @@
+      // not a registered client, do nothing
+      return false;
+    }
+
+    // external_correlation_id is simply the timestamp of this event,
+    // relative to profiling_start_time. i.e., it was computed as:
+    // external_correlation_id =
+    //      std::chrono::duration_cast<std::chrono::microseconds>(event_start_time - profiling_start_time).count()
+    //
+    // Because of the relative nature of the external_correlation_id, the same
+    // external_correlation_id can be reused across different clients, which then makes it
+    // impossible to recover the client from the external_correlation_id, which in turn
+    // makes it impossible to map events (which are tagged with external_correlation_id) to clients.
+    //
+    // To address these difficulties, we construct a new correlation_id (let's call it unique_cid)
+    // as follows:
+    // unique_cid =
+    //    external_correlation_id +
+    //    std::chrono::duration_cast<std::chrono::microseconds>(profiling_start_time.time_since_epoch()).count()
+    // now, unique_cid is monotonically increasing with time, so it can be used to reliably map events to clients.
+    //
+    // Of course, clients expect lists of events to be returned (on a call to Consume()), that are
+    // still keyed on the external_correlation_id that they've specified here, so we need to remember the
+    // offset to be subtracted
+    uint64_t offset = std::chrono::duration_cast<std::chrono::microseconds>(profiling_start_time.time_since_epoch()).count();
+    auto unique_cid = external_correlation_id + offset;
+    unique_correlation_id_to_client_offset_[unique_cid] = std::make_pair(client_handle, offset);
+    return this_as_derived->PushUniqueCorrelation(unique_cid);
+  }
+
+  void PopCorrelation(uint64_t& popped_external_correlation_id) {
+    auto this_as_derived = static_cast<TDerived*>(this);
+    std::lock_guard<std::mutex> lock(manager_instance_mutex_);
+    if (!logging_enabled_) {
+      return;
+    }
+    uint64_t unique_cid;
+    this_as_derived->PopUniqueCorrelation(unique_cid);
+    // lookup the offset and subtract it before returning popped_external_correlation_id to the client
+    auto client_it = unique_correlation_id_to_client_offset_.find(unique_cid);
+    if (client_it == unique_correlation_id_to_client_offset_.end()) {
+      popped_external_correlation_id = 0;
+      return;
+    }
+    popped_external_correlation_id = unique_cid - client_it->second.second;
+  }
+
+  void PopCorrelation() {
+    uint64_t unused;
+    PopCorrelation(unused);
+  }
+
+ protected:
+  GPUTracerManager() = default;
+
+#if 0
+  // Functional API to be implemented by subclasses
+  // Included here only for documentation purposes
+  bool OnStartLogging();
+  void OnStopLogging();
+  void ProcessActivityBuffers(const std::vector<ProfilerActivityBuffer>& buffers,
+                              const TimePoint& start_time);
+  bool PushUniqueCorrelation(uint64_t unique_cid);
```

Comment:
If possible, you shall opt for Status to indicate error state. But since it is not propagated through `GPUTracerManager`, I think keep it as is is also OK.

Reply:
Changed to return `void`: It isn't clear what higher level code can do if a basic functionality like this fails.